### PR TITLE
Network sync refactoring (part 5)

### PR DESCRIPTION
**Pull request by nazar-pc**
Another step towards https://github.com/paritytech/substrate/issues/8686

This addresses ugliness in configuration with factory method after #11412 as discussed in https://github.com/paritytech/substrate/pull/11412#discussion_r901736355

Not entirely happy with sync mode matching in multiple places, open for suggestions.

### CODE REVIEW

**Comment by nazar-pc** [comment]
Resolved merge conflicts after #11865.

@bkchr PTAL.
I have much bigger refactoring pending submission and due to the nature of changes long delays make it harder to resolve merge conflicts.

**Comment by nazar-pc** [comment]
@arkpar you were reviewing previous PR, maybe you can take a look?

**Comment by arkpar** [comment]
I'd suggest removing `sc_network::config::SyncMode` and simply using `sc_network_common::sync::SyncMode` everywhere.
Looks good otherwise.

File: client/service/src/builder.rs [hunk 1]
```diff
@@ -776,7 +778,18 @@ where
 		protocol_config
 	};
 
-	let max_parallel_downloads = config.network.max_parallel_downloads;
+	let chain_sync = ChainSync::new(
+		match config.network.sync_mode.clone() {
```
Comment (bkchr) [inline client/service/src/builder.rs:782]:
```suggestion
		match config.network.sync_mode {
```

File: client/network/test/src/lib.rs [hunk 2]
```diff
@@ -835,10 +835,25 @@ where
 			protocol_config
 		};
 
-		let max_parallel_downloads = network_config.max_parallel_downloads;
 		let block_announce_validator = config
 			.block_announce_validator
 			.unwrap_or_else(|| Box::new(DefaultBlockAnnounceValidator));
+		let chain_sync = ChainSync::new(
+			match network_config.sync_mode.clone() {
```
Comment (bkchr) [inline client/network/test/src/lib.rs:842]:
```suggestion
			match network_config.sync_mode {
```

File: client/network/src/service/tests.rs [hunk 3]
```diff
@@ -111,35 +111,36 @@ fn build_test_full_node(
 		protocol_config
 	};
 
-	let max_parallel_downloads = config.max_parallel_downloads;
+	let chain_sync = ChainSync::new(
+		match network_config.sync_mode.clone() {
```
Comment (bkchr) [inline client/network/src/service/tests.rs:115]:
```suggestion
		match network_config.sync_mode {
```

Comment (bkchr) [reply client/network/src/service/tests.rs:115]: (We should derive `Copy` for the type)

**Review by bkchr — APPROVED** [review]
As always, sorry for the delay!

**Review by arkpar — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**